provisioning/baremetal_webhook.go

```diff
@@ -0,0 +1,159 @@
+		return false, err
+	}
+
+	vw := newBaremetalOperatorWebhook(info.Namespace)
+	expectedGeneration := resourcemerge.ExpectedValidatingWebhooksConfiguration(validatingWebhookConfigurationName, info.ProvConfig.Status.Generations)
+	validatingWebhook, updated, err := resourceapply.ApplyValidatingWebhookConfiguration(info.Client.AdmissionregistrationV1(), info.EventRecorder, vw, expectedGeneration)
+	if err != nil {
+		err = errors.Wrap(err, "unable to create validatingwebhook configuration")
+		return false, err
+	}
+
+	if updated {
+		resourcemerge.SetValidatingWebhooksConfigurationGeneration(&info.ProvConfig.Status.Generations, validatingWebhook)
+	}
+
+	return updated, nil
+}
+
+// BaremetalWebhookDependenciesReady checks dependencies to enable Baremetal
+// Operator ValidatingWebhook.
+func BaremetalWebhookDependenciesReady(client osclientset.Interface) bool {
+	// Service CA operator will inject required certificates.
+	// If Service CA is not ready, ValidatingWebhook should not be enabled.
+	return serviceCAOperatorReady(client)
+}
+
+// DeleteValidatingWebhook deletes ValidatingWebhookConfiguration and
+// service resources.
+func DeleteValidatingWebhook(info *ProvisioningInfo) error {
+	err := client.IgnoreNotFound(info.Client.CoreV1().Services(info.Namespace).Delete(context.Background(), validatingWebhookService, metav1.DeleteOptions{}))
+	if err != nil {
+		return err
+	}
+
+	err = client.IgnoreNotFound(info.Client.AdmissionregistrationV1().
+		ValidatingWebhookConfigurations().
+		Delete(context.Background(), validatingWebhookConfigurationName, metav1.DeleteOptions{}))
+
+	return err
+}
+
+func newBaremetalOperatorWebhookService(targetNamespace string) *corev1.Service {
+	webhookPort, _ := strconv.ParseInt(baremetalWebhookPort, 10, 32) // #nosec
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      validatingWebhookService,
+			Namespace: targetNamespace,
+			Annotations: map[string]string{
+				"include.release.openshift.io/self-managed-high-availability": "true",
+				"include.release.openshift.io/single-node-developer":          "true",
+				"service.beta.openshift.io/serving-cert-secret-name":          baremetalWebhookSecretName,
+			},
+			Labels: map[string]string{
+				"k8s-app":    metal3AppName,
+				cboLabelName: stateService,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Type: corev1.ServiceTypeClusterIP,
+			Selector: map[string]string{
+				"k8s-app":    metal3AppName,
+				cboLabelName: stateService,
+			},
+			Ports: []corev1.ServicePort{
+				{
+					Name:       httpPortName,
+					Port:       validatingWebhookServiceHttpsPort,
+					TargetPort: intstr.FromInt(int(webhookPort)),
+				},
+			},
+		},
+	}
+}
+
+func newBaremetalOperatorWebhook(targetNamespace string) *admissionregistration.ValidatingWebhookConfiguration {
+	failurePolicy := admissionregistration.Fail
+	sideEffect := admissionregistration.SideEffectClassNone
+	instance := &admissionregistration.ValidatingWebhookConfiguration{
```

Comment:
While this looks fine, it also seems like this is something that could be applied directly as a manifest based on the upstream one from the BMO repo with some light kustomization, just as we do for the CRD. That advantage of doing that instead of doing it in the operator is (a) less code, but also (b) they can't get out of sync. WDYT?

Reply:
In upstream, we can safely apply webhook manifests because cluster-api ensures that cert-manager is running and healthy. In downstream, we rely on service-ca but there is no guarantee that it is healthy. Probably, that's why, CBO also uses code instead manifests as in here https://github.com/openshift/cluster-baremetal-operator/blob/af0136a0dba7f42f2a493ba7a09dc33f81d45c41/provisioning/webhook.go#L38. 

And I agree about less code and always up to date, but ValidatingWebhookConfiguration has very few fields that can be changeable once it is set.